frappe/utils/safe_exec.py

```diff
@@ -269,11 +269,33 @@ def get_hooks(hook=None, default=None, app_name=None):
 def read_sql(query, *args, **kwargs):
 	'''a wrapper for frappe.db.sql to allow reads'''
 	query = str(query)
-	if frappe.flags.in_safe_exec and not query.strip().lower().startswith('select'):
-		raise frappe.PermissionError('Only SELECT SQL allowed in scripting')
+	if frappe.flags.in_safe_exec:
+		check_safe_sql_query(query)
 	return frappe.db.sql(query, *args, **kwargs)
 
 
+def check_safe_sql_query(query: str, throw: bool = True) -> bool:
+	""" Check if SQL query is safe for running in restricted context.
+
+		Safe queries:
+			1. Read only 'select' or 'explain' queries
+			2. CTE on mariadb where writes are not allowed.
+	"""
+
+	query = query.strip().lower()
+	whitelisted_statements = ("select", "explain")
+
+	if (query.startswith(whitelisted_statements)
+		or (query.startswith("with") and frappe.db.db_type == "mariadb")):
```

Comment:
@gavindsouza @surajshetty3416 

should we instead parse the query and check if any DML statement is `non-select`? That will allow the use of CTE on both DB. 

https://sqlparse.readthedocs.io/en/latest/analyzing/  (this is already a dependency)

Reply:
I have a feeling that would add too much overhead without much gain - it doesn't handle SQLis for both our DBs effectively either. We could port some changes from https://github.com/frappe/frappe/pull/14920 though.

Reply:
lets go ahead with this for now then.